pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPushUtils.java

```diff
@@ -0,0 +1,349 @@
+        int attempts = 1;
+        if (spec.getPushJobSpec() != null && spec.getPushJobSpec().getPushAttempts() > 0) {
+          attempts = spec.getPushJobSpec().getPushAttempts();
+        }
+        long retryWaitMs = 1000L;
+        if (spec.getPushJobSpec() != null && spec.getPushJobSpec().getPushRetryIntervalMillis() > 0) {
+          retryWaitMs = spec.getPushJobSpec().getPushRetryIntervalMillis();
+        }
+        RetryPolicies.exponentialBackoffRetryPolicy(attempts, retryWaitMs, 5).attempt(() -> {
+          try (InputStream inputStream = fileSystem.open(tarFileURI)) {
+            SimpleHttpResponse response = FILE_UPLOAD_DOWNLOAD_CLIENT
+                .uploadSegment(FileUploadDownloadClient.getUploadSegmentURI(controllerURI), segmentName, inputStream,
+                    FileUploadDownloadClient.makeAuthHeader(spec.getAuthToken()),
+                    FileUploadDownloadClient.makeTableParam(tableName),
+                    FileUploadDownloadClient.DEFAULT_SOCKET_TIMEOUT_MS);
+            LOGGER.info("Response for pushing table {} segment {} to location {} - {}: {}", tableName, segmentName,
+                controllerURI, response.getStatusCode(), response.getResponse());
+            return true;
+          } catch (HttpErrorStatusException e) {
+            int statusCode = e.getStatusCode();
+            if (statusCode >= 500) {
+              // Temporary exception
+              LOGGER.warn("Caught temporary exception while pushing table: {} segment: {} to {}, will retry", tableName,
+                  segmentName, controllerURI, e);
+              return false;
+            } else {
+              // Permanent exception
+              LOGGER
+                  .error("Caught permanent exception while pushing table: {} segment: {} to {}, won't retry", tableName,
+                      segmentName, controllerURI, e);
+              throw e;
+            }
+          } finally {
+            if (cleanUpOutputDir) {
+              fileSystem.delete(tarFileURI, true);
+            }
+          }
+        });
+      }
+    }
+  }
+
+  public static void sendSegmentUris(SegmentGenerationJobSpec spec, List<String> segmentUris)
+      throws RetriableOperationException, AttemptsExceededException {
+    String tableName = spec.getTableSpec().getTableName();
+    LOGGER.info("Start sending table {} segment URIs: {} to locations: {}", tableName,
+        Arrays.toString(segmentUris.subList(0, Math.min(5, segmentUris.size())).toArray()),
+        Arrays.toString(spec.getPinotClusterSpecs()));
+    for (String segmentUri : segmentUris) {
+      URI segmentURI = URI.create(segmentUri);
+      PinotFS outputDirFS = PinotFSFactory.create(segmentURI.getScheme());
+      for (PinotClusterSpec pinotClusterSpec : spec.getPinotClusterSpecs()) {
+        URI controllerURI;
+        try {
+          controllerURI = new URI(pinotClusterSpec.getControllerURI());
+        } catch (URISyntaxException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)